prescient/simulator/config.py

```diff
@@ -0,0 +1,407 @@
+#  ___________________________________________________________________________
+#
+#  Prescient
+#  Copyright 2020 National Technology & Engineering Solutions of Sandia, LLC
+#  (NTESS). Under the terms of Contract DE-NA0003525 with NTESS, the U.S.
+#  Government retains certain rights in this software.
+#  This software is distributed under the Revised BSD License.
+#  ___________________________________________________________________________
+
+####################################################
+#                 Config                           #
+####################################################
+
+from __future__ import annotations
+import sys
+from typing import List
+
+from argparse import ArgumentParser
+from pyutilib.misc import import_file
+from pyomo.common.config import (ConfigDict,
+                                 ConfigValue,
+                                 ConfigList,
+                                 In,
+                                 PositiveInt,
+                                 NonNegativeInt,
+                                 PositiveFloat,
+                                 NonNegativeFloat,
+                                )
+
+import prescient.plugins
+
+prescient_persistent_solvers = ("cplex", "gurobi", "xpress")
+prescient_solvers = [ s+sa for sa in ["", "_direct", "_persistent"] for s in prescient_persistent_solvers ]
+prescient_solvers += ["cbc", "glpk"]
+
+def parse_args(args: List[str]) -> ConfigDict:
+    parser = _construct_options_parser()
+    args = parser.parse_args(args=args)
+    PrescientConfig.import_argparse(args)
+    return PrescientConfig
+
+def _construct_options_parser() -> ArgumentParser:
+    '''
+    Make a new parser that can parse standard and custom command line options.
+
+    Custom options are provided by plugin modules. Plugins are specified
+    as command-line arguments "--plugin=<module name>", where <module name>
+    refers to a python module. The plugin may provide new command line
+    options by calling "prescient.plugins.add_custom_commandline_option"
+    when loaded.
+    '''
+
+    # To support the ability to add new command line options to the line that
+    # is currently being parsed, we convert a standard ArgumentParser into a
+    # two-pass parser by replacing the parser's parse_args method with a
+    # modified version.  In the modified parse_args, a first pass through the
+    # command line finds any --plugin arguments and allows the plugin to 
+    # add new options to the parser.  The second pass does a full parse of
+    # the command line using the parser's original parse_args method.
```

Comment:
One issue I ran into is that default values didn't get set if you added new options mid-parse. This was back when we were using OptionParser instead of ArgumentParser, but you should check for this.